lib/routes/utils/signin.js

```diff
@@ -0,0 +1,485 @@
+          throw e
+        }
+        return request.emitMetricsEvent('account.login.blocked').then(() => {
+          // If this customs error cannot be bypassed with email confirmation,
+          // throw it straight back to the caller.
+          var verificationMethod = e.output.payload.verificationMethod
+          if (verificationMethod !== 'email-captcha' || ! request.payload.unblockCode) {
+            throw e
+          }
+          // Check for a valid unblockCode, to allow the request to proceed.
+          // This requires that we load the accountRecord to learn the uid.
+          const unblockCode = request.payload.unblockCode.toUpperCase()
+          return db.accountRecord(email).then(result => {
+            accountRecord = result
+            return db.consumeUnblockCode(accountRecord.uid, unblockCode).then(code => {
+              if (Date.now() - code.createdAt > unblockCodeLifetime) {
+                log.info({
+                  op: 'Account.login.unblockCode.expired',
+                  uid: accountRecord.uid
+                })
+                throw error.invalidUnblockCode()
+              }
+            }).then(() => {
+              didSigninUnblock = true
+              return request.emitMetricsEvent('account.login.confirmedUnblockCode')
+            }).catch((e) => {
+              if (e.errno !== error.ERRNO.INVALID_UNBLOCK_CODE) {
+                throw e
+              }
+              return request.emitMetricsEvent('account.login.invalidUnblockCode').then(() => {
+                throw e
+              })
+            })
+          })
+        })
+      }).then(() => {
+        // If we didn't load it above while checking unblock codes,
+        // it's now safe to load the account record from the db.
+        if (! accountRecord) {
+          return db.accountRecord(email).then(result => {
+            accountRecord = result
+          })
+        }
+      }).then(() => {
+        return { accountRecord, didSigninUnblock }
+      }).catch((e) => {
+        // Some errors need to be flagged with customs.
+        if (e.errno === error.ERRNO.INVALID_UNBLOCK_CODE || e.errno === error.ERRNO.ACCOUNT_UNKNOWN) {
+          customs.flag(request.app.clientAddress, {
+            email: email,
+            errno: e.errno
+          })
+        }
+        // For any error other than INVALID_UNBLOCK_CODE, hide it behind the original customs error.
+        // This prevents us from accidentally leaking additional info to a caller that's been
+        // blocked, including e.g. whether or not the target account exists.
+        if (originalError && e.errno !== error.ERRNO.INVALID_UNBLOCK_CODE) {
+          throw originalError
+        }
+        throw e
+      })
+    },
+
+    /**
+    * Send all the various notifications that result from a new signin.
+    * This includes emailing the user, logging metrics events, and
+    * notifying attached services.
+    */
```

Comment:
Here, the stars do not align.

(that was an indentation pun, in case it's not clear)